internal/servicelog/formatter.go

```diff
@@ -151,20 +158,70 @@ func (w *TrimWriter) Write(p []byte) (int, error) {
 	}
 }
 
-func (w *TrimWriter) write(p []byte) error {
+func (w *TrimWriter) write(line []byte) error {
+	const maxDetectFailures = 10
+	if w.layout == "" && w.detectFailures < maxDetectFailures {
+		w.layout = detectLayout(line)
+		if w.layout == "" {
+			w.detectFailures++
+		}
+	}
+
 	start := 0
-	loc := w.re.FindIndex(p)
-	if loc != nil {
-		start = loc[1]
+	if w.layout != "" {
+		prefix := timeLayoutPrefix(w.layout, line)
+
+		// TODO: do we want to just trust the field splitting and not try to
+		// parse every prefix?  This could potentially cause us to discard
+		// non-timestamp content from log lines.
+		_, err := time.Parse(w.layout, prefix)
+		if err == nil {
+			start = len(prefix)
+		} else {
+			// TODO: should this be an error?
+		}
 	}
 
-	p = p[start:]
-	for len(p) > 0 {
-		n, err := w.dest.Write(p)
-		p = p[n:]
+	line = line[start:]
+	for len(line) > 0 {
+		n, err := w.dest.Write(line)
+		line = line[n:]
 		if err != nil {
 			return err
 		}
 	}
 	return nil
 }
+
+func timeLayoutPrefix(layout string, line []byte) string {
+	if layout == "" {
+		return ""
+	}
+
+	nfields := len(strings.Split(layout, " "))
+	sep := []byte{' '}
+	fields := bytes.SplitN(line, sep, nfields)
+	size := 0
+	if len(fields) >= nfields {
+		// last layout char must be a space otherwise we can't detect the
+		// prefix using whitespace split/fields approach. This causes the last
+		// split field to be an empty string in the layout which we don't want
+		// to map to keeping any log line content - so we exclude
+		// this by reducing the number of fields we keep by one.
+		for _, field := range fields[:nfields-1] {
+			size += len(field) + len(sep)
+		}
+	}
+	return string(line[:size])
```

Comment:
Not sure how much difference it takes compared to the actual time parsing, but ideally this would be done with just slicing the original byte slice to avoid to-from string allocations.